drivers: serial: uart_mcux_lpuart: add DTR, DSR, and DCD signals

DerekSnell opened the pull request:
Add support for additional line control signals.  Improved configuring uart_config_flow_control options.  

Removed macro UART_LINE_CTRL_ENABLE because some line control signals in the LPUART do not depend on FSL_FEATURE_LPUART_HAS_MODEM_SUPPORT.  Added LPUART_HAS_MODEM instead.

decsny commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -1016,22 +1016,43 @@ static int mcux_lpuart_configure_basic(const struct device *dev, const struct ua
+#endif
+
+	case UART_CFG_FLOW_CTRL_DTR_DSR:
+#if defined(FSL_FEATURE_LPUART_HAS_MCR) && FSL_FEATURE_LPUART_HAS_MCR
```
make macro #define for this expression since it shows up many times

DerekSnell replied:
done, created `LPUART_HAS_MCR`

decsny commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -30,10 +30,10 @@ LOG_MODULE_REGISTER(uart_mcux_lpuart, LOG_LEVEL_ERR);
+/* Helper macros */
+#if defined(FSL_FEATURE_LPUART_HAS_MODEM_SUPPORT) && \
+	FSL_FEATURE_LPUART_HAS_MODEM_SUPPORT
+#define LPUART_HAS_MODEM
```
let's change this to be defined to 1, so that we can use IS_ENABLED

DerekSnell replied:
done, thank you

decsny commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -1016,22 +1016,43 @@ static int mcux_lpuart_configure_basic(const struct device *dev, const struct ua
+		break;
+#else
+		return -ENOTSUP;
+#endif
```
instead of a bunch of cases with a bunch of #if #else, just check like

```
if (!IS_ENABLED(LPUART_HAS_MODEM) && cfg->flow_ctrl != UART_CONFIG_FLOW_CTRL_NONE) {
  return -ENOTSUP;
}
```

at the start of the function, similar with the MCR expression using another macro like suggested above

and consider making helper functions

decsny commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -1199,6 +1220,21 @@ static void mcux_lpuart_line_ctrl_set_rts(const struct mcux_lpuart_config *confi
 		config->base->MODIR |= (LPUART_MODIR_TXRTSPOL_MASK | LPUART_MODIR_TXRTSE_MASK);
 	}
 }
+#endif /* LPUART_HAS_MODEM */
```
make an #else case taht does like,

#define mcux_lpuart_line_ctrl_set_rts(...)

then you can remove all the ifdefs below, similar for other functions which may or may not be defined, it makes the rest of the code more readable

DerekSnell replied:
very helpful, thank you.  I used this a couple times for the helper functions

decsny commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -1208,11 +1244,22 @@ static int mcux_lpuart_line_ctrl_set(const struct device *dev,
 
 	switch (ctrl) {
 	case UART_LINE_CTRL_RTS:
+#ifdef LPUART_HAS_MODEM
 		/* Disable Transmitter and Receiver */
 		config->base->CTRL &= ~(LPUART_CTRL_TE_MASK | LPUART_CTRL_RE_MASK);
 
 		mcux_lpuart_line_ctrl_set_rts(config, val);
+#else
+		ret = -ENOTSUP;
+#endif
+		break;
 
+	case UART_LINE_CTRL_DTR:
+#if defined(FSL_FEATURE_LPUART_HAS_MCR) && FSL_FEATURE_LPUART_HAS_MCR
+		mcux_lpuart_set_dtr(config, val);
+#else
+		ret = -ENOTSUP;
+#endif
```
same suggestion about making an initial check instead of doing the same #ifdefs over and over

decsny reviewed (dismissed):
good addition but the code is so messy, so for maintenance reasons I can't accept it in current state, there are many opportunities for cleanup which I put in the review

DerekSnell commented:
> good addition but the code is so messy, so for maintenance reasons I can't accept it in current state, there are many opportunities for cleanup which I put in the review

Thank you @decsny for your feedback.  My code was messy, and your suggestions do improve the readability.  Please let me know how this looks to you know.  Thanks

decsny commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -970,9 +974,47 @@ static void mcux_lpuart_isr(const struct device *dev)
+static int mcux_lpuart_config_flowctrl(uint8_t flow_ctrl, lpuart_config_t *uart_config)
+{
+	switch (flow_ctrl) {
+	case UART_CFG_FLOW_CTRL_NONE:
+		uart_config->enableTxCTS = false;
+		uart_config->enableRxRTS = false;
+		break;
+
+	case UART_CFG_FLOW_CTRL_RS485:
+		uart_config->enableTxCTS = false;
+		uart_config->enableRxRTS = false;
+		break;
```
since these are the same you can use case fallthrough to eliminate a couple lines

DerekSnell replied:
Good catch, @decsny .  Updated

decsny commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -970,9 +974,47 @@ static void mcux_lpuart_isr(const struct device *dev)
+#else
+static int mcux_lpuart_config_flowctrl(uint8_t flow_ctrl, lpuart_config_t *uart_config)
+{
+	if (flow_ctrl == UART_CFG_FLOW_CTRL_NONE) {
+		return 0;
+	}
+
+	return -ENOTSUP;
```
very nit but would prefer to have the

```
if (flow_ctrl != ...NONE) {
  return -ENOTSUP
}

return 0;
```
just because thats the normal paradigm is to check errors and then last return 0. right now not a big deal but if function gets bigger this is easier to edit

DerekSnell replied:
makes sense, updated

decsny reviewed (commented):
this is much nicer and more reviewable too, thank you for doing that cleanup, I just have a couple minor nits left but they are not blocking

Commit:
drivers: serial: uart_mcux_lpuart: add DTR, DSR, and DCD signals

Add support for additional line control signals.  Improved configuring
uart_config_flow_control options.

Signed-off-by: Derek Snell <derek.snell@nxp.com>

DerekSnell commented:
@abhinavnxp and @nirav-agrawal , FYI

Since you are using HW flow control with modules and the line control APIs in LPUART, be aware of this update to the driver.  Feel free to test out this update.  Best regards
